### PR TITLE
WIP: Add watch cache size test

### PR DESCRIPTION
**Pull request by sttts**

### CODE REVIEW

**Comment by sttts** [comment]
@ncdc @deads2k what do I miss that the second watch call passes without error?

**Comment by deads2k** [comment]
> @ncdc @deads2k what do I miss that the second watch call passes without error?

@liggitt Don't we fall back to trying to pull the watch from etcd so if etcd still has the resourceVersion we still work?

**Comment by liggitt** [comment]
> @liggitt Don't we fall back to trying to pull the watch from etcd so if etcd still has the resourceVersion we still work?

I don't remember that happening

**Comment by sttts** [comment]
Good point, didn't know that. Will check.

**Comment by deads2k** [comment]
> I don't remember that happening

Really?  It's either/or? That seems weirdly strict.

File: test/integration/watch_cache_test.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	}
+	startVersion, err := strconv.Atoi(defaultNS.ResourceVersion)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	for i := 0; i < expectedCacheSize+10; i++ {
+		defaultNS, err = patchClient.Namespaces().Patch("default", types.StrategicMergePatchType,
+			[]byte(fmt.Sprintf(`{"metadata":{"labels":{"test":"%d"}}}`, i)))
+		if err != nil {
+			t.Fatalf("unexpected patch error: %v", err)
```
Comment (deads2k) [inline test/integration/watch_cache_test.go]: You're making a high contention field and racing the cache.  I think you'll be ok if you mutate the label too.

File: test/integration/watch_cache_test.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		t.Fatalf("unexpected error converting the resource version: %v", err)
+	}
+	w, err := client.Core().Namespaces().Watch(metav1.ListOptions{ResourceVersion: strconv.Itoa(rv - (expectedCacheSize-counterExampleCacheSize)/2)})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	defer w.Stop()
+
+	// try watch with an version that is too old
+	rv, err = strconv.Atoi(defaultNS.ResourceVersion)
```
Comment (deads2k) [inline test/integration/watch_cache_test.go]: why recompute this?

File: test/integration/watch_cache_test.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	if err != nil {
+		t.Fatalf("unexpected error converting the resource version: %v", err)
+	}
+	w, err = client.Core().Namespaces().Watch(metav1.ListOptions{ResourceVersion: strconv.Itoa(startVersion - 1)})
```
Comment (deads2k) [inline test/integration/watch_cache_test.go]: Alright, here's a theory. Our cache hasn't caught up yet?  Try a get with `GetOptions` set to your latest resource version before starting the watch.

File: test/integration/watch_cache_test.go (hunk 1 above)
Comment (sttts) [inline test/integration/watch_cache_test.go]: in this integration test context it should be fine, shouldn't it?

File: test/integration/watch_cache_test.go (hunk 2 above)
Comment (sttts) [inline test/integration/watch_cache_test.go]: copy&paste error

File: test/integration/watch_cache_test.go (hunk 3 above)
Comment (sttts) [inline test/integration/watch_cache_test.go]: A get with an explicit version doesn't help.

File: test/integration/watch_cache_test.go (hunk 1 above)
Comment (deads2k) [inline test/integration/watch_cache_test.go]: > in this integration test context it should be fine, shouldn't it?

I don't think so.  You can get a conflict error and end up fatal-ing your test.

**Pull request closed without merging**